lib/watcher-middleware.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+var url = require('url');
+var path = require('path');
+
+var setResponseHeaders = require('./utils/response-header');
+var errorHandler = require('./utils/error-handler');
+
+module.exports = function watcherMiddleware(watcher, options) {
+  options = options || {};
+
+  if (!options.hasOwnProperty('autoIndex')) {
+    // set autoIndex to be true if not provided
+    options.autoIndex = true;
+  }
+
+  return function(request, response, next) {
+    watcher.then(function(hash) {
+      var outputPath = path.normalize(hash.directory);
+
+      // set the x-broccoli header containing per request info used by the broccoli-middleware
+      var urlToBeServed = request.url;
+      var urlObj = url.parse(urlToBeServed);
+      var filename = path.join(outputPath, decodeURIComponent(urlObj.pathname));
+      var broccoliInfo = {
+        'url': urlToBeServed,
+        'outputPath': outputPath,
+        'filename': filename,
+      };
+      request.headers['x-broccoli'] = broccoliInfo;
+
+      var updatedFileName = setResponseHeaders(request, response, next, {
```

Comment:
> The purpose of the watcher middleware was to set all the generic headers and the serve file asset middleware to actually just serve from disk

I would have imagined the purpose of the watcher middleware to just be the error handling and to block the request until the watcher has finished.

> The reason being: Imagine usecase where someone wants to insert a middleware between the watcher and serve asset middleware in order to serve assets themselves. If the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers.

So a couple of things:

1. If someone else is serving assets presumably there's something different about them, which is why they can't just fall back to the asset serving middleware.  This means that at a minimum headers like [Content-Length](https://github.com/ember-cli/broccoli-middleware/blob/aac8215f5184b226c51e964313c5c49460cf0da1/lib/utils/response-header.js#L116) and [Last-Modified](https://github.com/ember-cli/broccoli-middleware/blob/aac8215f5184b226c51e964313c5c49460cf0da1/lib/utils/response-header.js#L101) are likely to be wrong.

2. It seems weird that they can only sometimes rely on the default headers.  The headers set in `setResponseHeaders` depend on being able to actually find the file on disk.  Someone who wants to write a middleware between the watcher and asset serving may or may not actually be serving assets from disk or not.  The reason this isn't a problem with the asset serving middleware is that serving files found on disk takes precedence over the proxy, but for someone who wanted to insert their middleware *before* asset serving, they would want their middleware to take precedence.

> if the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers.

Maybe we could move that logic into `lib/utils/serve-asset.js` and make it public?  This way it would be still be pretty easy to write a middleware that wanted to serve a static asset just like serve assets, but whose url -> file on disk algorithm was different (eg as with fastboot which wants `/` -> `/index.fastboot.html` instead of `/` -> `/index.html` or something similar)